python/ray/tune/schedulers/resource_changing_scheduler.py

```diff
@@ -14,130 +16,374 @@
-                t.placement_group_factory.required_resources.get("CPU", 0),
-                t.placement_group_factory.required_resources.get("GPU", 0),
+        return upper_cpu_limit, upper_gpu_limit
+
+    def _modify_upper_limits_with_increase_by_times(self) -> None:
+        if self.increase_by and self.increase_by_times > 0:
+            required_cpus = self.increase_by.get("CPU", 0)
+            required_gpus = self.increase_by.get("GPU", 0)
+            self.upper_cpu_limit = min(
+                self.upper_cpu_limit,
+                self.min_cpu + required_cpus * self.increase_by_times,
+            )
+            self.upper_gpu_limit = min(
+                self.upper_gpu_limit,
+                self.min_gpu + required_gpus * self.increase_by_times,
             )
 
+    def _modify_lower_limits_with_increase_by_times(self) -> None:
+        if self.increase_by:
+            required_cpus = self.increase_by.get("CPU", 0)
+            required_gpus = self.increase_by.get("GPU", 0)
+            if required_cpus and required_gpus:
+                multiplier = min(
+                    self.free_cpus // required_cpus, self.free_gpus // required_gpus
+                )
+            elif required_gpus:
+                multiplier = self.free_gpus // required_gpus
+            else:
+                multiplier = self.free_cpus // required_cpus
+            multiplier = max(multiplier, 0)
+            self.free_cpus = multiplier * required_cpus
+            self.free_gpus = multiplier * required_gpus
+
+    def _get_used_cpus_and_gpus(self, t: Trial):
+        """Function to check how many CPUs and GPUs a trial is using
+        currently"""
+        return (
+            t.placement_group_factory.required_resources.get("CPU", 0),
+            t.placement_group_factory.required_resources.get("GPU", 0),
+        )
+
+    def __call__(
+        self,
+        trial_runner: "trial_runner.TrialRunner",
+        trial: Trial,
+        result: Dict[str, Any],
+        scheduler: "ResourceChangingScheduler",
+    ) -> Union[None, PlacementGroupFactory]:
+        """Run resource allocation logic.
+
+        Returns a new ``PlacementGroupFactory`` with updated
+        resource requirements, or None. If the returned
+        ``PlacementGroupFactory`` is equal by value to the one the
+        trial has currently, the scheduler will skip the update process
+        internally (same with None).
+
+        Args:
+            trial_runner (TrialRunner): Trial runner for this Tune run.
+                Can be used to obtain information about other trials.
+            trial (Trial): The trial to allocate new resources to.
+            result (Dict[str, Any]): The latest results of trial.
+            scheduler (ResourceChangingScheduler): The scheduler calling
+                the function.
+        """
+        self.trial_runner = trial_runner
+        self.trial = trial
+        self.result = result
+        self.scheduler = scheduler
```

Comment:
I could be missing it, but how is `self.scheduler` used?

Reply:
It is not used here, but is left for user defined subclasses.

Reply:
Is there an example of how a user may use scheduler in this user defined subclass? It's a bit abstract for me to imagine right now.

Reply:
It can be used to eg. persist some state between calls (like the last resized trial), since the DistributedResource callable is not supposed to hold any.